### PR TITLE
generalize the distributed operations

### PR DESCRIPTION
**Pull request by exaexa**
So this vastly simplifies all the distributed handling by converting the operations to a single base.

- instead of references, DArrays and whatnot, there is just a single type of syntactical reference to data
- there is a framework for saving/retrieving local data to nodes, mapped both atop the DistributedArrays and the JLS file distribution system
- the new framework supports transform + map/reduce, which is a nice abstraction for all operations we do. For example, the distributed call of SOM epoch now looks like this:
```julia
distributed_mapreduce(
        dataVal,
        (x) -> doEpoch(x, codes, tree),        #map step (runs a singlethreaded epoch on the slice)
        ((n1,d1), (n2,d2)) -> (n1+n2, d1+d2),   #reduce step (sums the numerators and denominators)
        workers)
```
  As a side effect, the code is now much shorter.
- the framework is most useful for analyzing really _huge_ data by hand. E.g. finding mean + scaling is now one call to `distributed_mapreduce` to get the means+sdevs, and one call to `distributed_transform` to apply the scaling.

GigaSOM ops on DistributedArrays are working well now; the tests for the JLS file slices are not fixed yet.

### CODE REVIEW

File: src/core.jl [hunk 1]
```diff
@@ -208,92 +92,61 @@ function trainGigaSOM(som::Som, train::DataFrame;
...
+    # get the SOM neighborhood distances
     dm = distMatrix(som.grid, som.toroidal)
 
     codes = som.codes
 
-    nWorkers = nprocs()
-    dTrain = distribute(train)
-
     for j in 1:epochs
+        @info "Epoch $j..."
```
Comment (laurentheirendt) [inline src/core.jl]: Maybe here could also be a `printLevel` so that the entire output could be muted at once (similar to the i/o splitting).

Comment (exaexa) [reply src/core.jl]: Is there some standard method for that or should we make one? (In C I'm usually just dumping stuff to `stderr` and everyone is welcome to ignore it with a redirect :D)

Comment (exaexa) [reply src/core.jl]: EDIT: there's apparently @debug and standardized loglevels; I'll use that.

Comment (laurentheirendt) [reply src/core.jl]: cool - sure, why not. maybe let's do another PR just for that

File: src/core.jl [hunk 2]
```diff
@@ -77,6 +24,7 @@ function initGigaSOM(train, xdim, ydim = xdim;
     numCodes = xdim * ydim
 
     # normalise training data:
+    # TODO: is this needed?
```
Comment (oHunewald) [inline src/core.jl]: not for our FCS data, but one could think of future projects with different type of data

**Comment by laurentheirendt** [comment]
As this is a PR with many changes, we need to make sure that the documentation builds correctly. Can you confirm, @exaexa ?

**Comment by exaexa** [comment]
@laurentheirendt It does build correctly, in the last commit I fixed the references to the new structure so that it makes sense.

I'm unsure about the actual "user guide" though -- will read through it and try to fix invalid stuff.

**Review by oHunewald — APPROVED** [review]

**Comment by laurentheirendt** [comment]
great job @exaexa 🎉 👍

**Pull request merged**